pkg/capacityscheduling/capacity_scheduling.go

```diff
@@ -0,0 +1,816 @@
+				Pods:             pods,
+				NumPDBViolations: int64(numPDBViolations),
+			}
+			c := candidate{
+				victims: &victims,
+				name:    nodeInfoCopy.Node().Name,
+			}
+			candidates = append(candidates, &c)
+			resultLock.Unlock()
+		}
+	}
+	pluginsutil.Until(ctx, len(potentialNodes), checkNode)
+	return candidates
+}
+
+// nodesWherePreemptionMightHelp returns a list of nodes with failed predicates
+// that may be satisfied by removing pods from the node.
+func nodesWherePreemptionMightHelp(nodes []*framework.NodeInfo, m framework.NodeToStatusMap) []*framework.NodeInfo {
+	var potentialNodes []*framework.NodeInfo
+	for _, node := range nodes {
+		name := node.Node().Name
+		// We reply on the status by each plugin - 'Unschedulable' or 'UnschedulableAndUnresolvable'
+		// to determine whether preemption may help or not on the node.
+		if m[name].Code() == framework.UnschedulableAndUnresolvable {
+			continue
+		}
+		potentialNodes = append(potentialNodes, node)
+	}
+	return potentialNodes
+}
+
+func selectVictimsOnNode(
+	ctx context.Context,
+	ph framework.PreemptHandle,
+	state *framework.CycleState,
+	pod *v1.Pod,
+	nodeInfo *framework.NodeInfo,
+	pdbs []*policy.PodDisruptionBudget,
+) ([]*v1.Pod, int, bool) {
+	elasticQuotaSnapshotState, err := getElasticQuotaSnapshotState(state)
+	if err != nil {
+		klog.Errorf("error reading %q from cycleState: %v", ElasticQuotaSnapshotKey, err)
+		return nil, 0, false
+	}
+
+	preFilterState, err := getPreFilterState(state)
+	if err != nil {
+		klog.Errorf("error reading %q from cycleState: %v", preFilterStateKey, err)
+		return nil, 0, false
+	}
+
+	var potentialVictims []*v1.Pod
+	var moreThanMinWithPreemptor bool
+
+	removePod := func(rp *v1.Pod) error {
+		if err := nodeInfo.RemovePod(rp); err != nil {
+			return err
+		}
+		status := ph.RunPreFilterExtensionRemovePod(ctx, state, pod, rp, nodeInfo)
+		if !status.IsSuccess() {
+			return status.AsError()
+		}
+		return nil
+	}
+
+	addPod := func(ap *v1.Pod) error {
+		nodeInfo.AddPod(ap)
+		status := ph.RunPreFilterExtensionAddPod(ctx, state, pod, ap, nodeInfo)
+		if !status.IsSuccess() {
+			return status.AsError()
+		}
+		return nil
+	}
+
+	podPriority := podutil.GetPodPriority(pod)
+	preemptorElasticQuotaInfo, preemptorWithElasticQuota := elasticQuotaSnapshotState.elasticQuotaInfos[pod.Namespace]
+
+	// Check if there is elastic quota in the preemptor's namespace.
+	if preemptorWithElasticQuota {
+		moreThanMinWithPreemptor = preemptorElasticQuotaInfo.overUsed(preFilterState.Resource, preemptorElasticQuotaInfo.Min, resourceAdd)
+	}
+
+	// sort the pods in node by the priority class
+	sort.Slice(nodeInfo.Pods, func(i, j int) bool { return !util.MoreImportantPod(nodeInfo.Pods[i].Pod, nodeInfo.Pods[j].Pod) })
+	for _, p := range nodeInfo.Pods {
+		pElasticQuotaInfo, pWithElasticQuota := elasticQuotaSnapshotState.elasticQuotaInfos[p.Pod.Namespace]
+		// Preemption only happens between in the same type of namespace: with or without.
```

Comment:
What if the pod gets unschedulabe is just due to regular constraints, and preemptorWithElasticQuota happens to not equal to pWithElasticQuota? I recalled that you expect this plugin to replace (instead of co-exist with) defaultpreemption plugin.

Reply:
If the pods get unschedulabe because of regular constraints，it will preempt other pods which belongs to ElasticQuota which eq.used > eq.min.

Reply:
> it will preempt other pods which belongs to ElasticQuota which eq.used > eq.min.

What if preempting RQ's managed pods isn't helpful - for example, a pod subject to RQ, but failed due to some regular constraints like PodAffinity. We may need to increase the searching scope of regular pods as we don't know the exact reason why it's judged as "unschedulable".

Reply:
If preemptorWithElasticQuota happens to not equal to pWithElasticQuota, it means one of preemptor or p doesn't belong to some EQ.  So the p can't be selected as victim. 
This is mainly to prevent mutual influence between the two sides. As I see, The Priority Class in the same EQ or tenant is comparable. The priority Class in the different EQ or tenant isn't comparable. This is the same here.

Reply:
Note that k8s doesn't have an official API called tenant yet. So if the implementation is assuming tenant == namespace, it should be documented well, otherwise the behavior would really confuse users.